README.md

```diff
@@ -502,6 +502,30 @@ When perfection is achieved and there is nothing left to `delete`, you can do th
 delete delete!
 ```
 
+## Forget
+
+You can forget variables which were declared, by using the `forget` keyword.
+
+```java
+const const name = "John Doe"!
+forget name! // Name has been forgotten
+print(name)! // Error/gibberish 
+```
+
+`forget` keyword works on variables declared with `const` and `const const`, but not variables declared with `const const const`
+
+```java
+const const const string = "C is the BEST LANGUAGE!!!"!
+forget string! // Does not work because const const const is immune to forget.
+print(string)! // Works
+```
+
+The forget keyword is good for explicit variable shadowing, better than amateur languages like rust which shadows variables implicitly.
+
+**Note**: When referencing a forgotten variable, it is not necessary that an error will be thrown, the application may return any random string like "Johny Doe", "John douglas" etc..
```

Comment:
How does C know the variable's type? I thought it forgot the variable?

Reply:
Wouldn't it make more sense for `name` to just be interpreted as a string after being forgotten, as strings can be declared without quotes? Sort of like the `noop` "keyword"

Reply:
True, that's actually in the [specs for strings](https://github.com/TodePond/C#strings).

Reply:
So I suppose a good way to characterize the `forget` keyword would be just a way of removing an identifier from scope. I quite like that honestly, as the `delete` keyword seems only geared to literals and primitives expressible as literals

Reply:
Python (for example) does this by using `del`. `delete` also removes keywords from the global scope. I currently don't see any problem with extending the scope of the `delete` keyword.

Reply:
Considering this, I actually do see one issue with merging the forget functionality into the delete keyword now. Consider this code:

```java
const const foo = 3!
delete foo!
```

In this case, should the delete keyword remove the `foo` identifier from the scope or remove the `3` literal from existence? Existing delete functionality makes it make more sense to do the latter, but the forget functionality dictates that the former be done. Merging functionality would make dynamic primitive deletion much more difficult in practice.

Maybe we can just say the forget keyword deletes reference identifiers from scope and extend the delete keyword to simply delete the target of those references (the objects) from the store?

Reply:
As far as I understand, `delete` deletes identifiers and stuff, but not any values. `forget` (as specified here) only deletes the value, but not the identifier. In this manner, to completely remove a variable from existence, you'd need to write something like `forget delete x`.

Reply:
Hmm, I think I'm interpreting the specification for `delete` a bit differently. Given that the only examples of usages of delete are on literals and keywords, I was taking it as exclusively deleting the values while leaving identifiers be. I'm realizing there isn't actually an example of delete being used on an identifier, so that's probably something which should be added to clear things up.

The way I'm reading it, deleting a variable from existence should be accomplishable with just a `forget x`, I would think, if you're willing to just settle for the garbage collector getting rid of the object. You could be more thorough with a `delete x` to get rid of the value (at runtime we could say, maybe x is set to a `deleted` pointer, sort of like a `null` pointer but better for error handling) and then a `forget x` to get rid of the identifier.

Reply:
You know `forget delete x` would actually be a really nice shorthand, for those two lines, now that I think about it.

Reply:
You're assuming a garbage collector, but it's nowhere defined how memory is managed. It could also be based on reference counting or explicit memory management. Nevertheless, your statement also holds for any kind of these, but I'd assume that it would have a temporal impact on the `forget` behaviour (i.e., values are forgotten eventually, but not immediately in case of a GC).

The difference between `forget` and `delete` as of right now is that `delete` will cause a compiler error, and `forget` will cause non-deterministic behaviour when using the forgotten variable.

Reply:
You're right, there isn't a memory management strategy defined and I shouldn't jump to that assumption.

I do see what you're saying. I just think `delete` has such a lack of clarity on how `delete identifier` behaves, and whether just the target of `identifier` or both target and identifier are considered "deleted" and unusable afterwards.

Reply:
If `delete` would remove the contents of a deleted identifier, `delete delete` would be a contradiction, as it would also delete the special error handling code for deleted keywords. Essentially, `delete` currently only blocks access to a keyword, but doesn't delete anything. In this case, `forget delete! delete delete!` would not work, as `delete` would have forgotten its behaviour when executing the second statement. And `delete delete! forget delete!` would also not work, as the second statement would cause an error accessing the deleted `delete`. To just remove an identifier, but not its contents, a third keyword must be introduced.